doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+between hosts and managers.
+
+For the purpose of this decision, we will assume that two versions of
+`mediacreation` can exist in the same process without problematic
+namespace collisions, although this remains a problem to be solved.
+
+(Note: Practically, this may mean that we end up strongly recommending
+ vendoring mediacreation into manager plugins, especially for python
+ libraries which can't do C++ tricks with private symbols.)
+
+## Schema version, package version
+
+The schema version (the version number in the `.yml`) and the package
+version (the `openassetio-mediacreation` library version) are separate.
+
+It is possible that updates to `openassetio` or `openassetio-traitgen`
+could cause breakages that necessitate a major version update to the
+`openassetio-mediacreation` package.
+
+We are assuming that this does **not** imply a schema version update.
+
+
+# What to Version
+In considering how to version the openassetio traits and specifications,
+two options immediately present themselves. Version every trait and
+specification individually, or version the entire schema as a whole.
+
+## Version every trait
+Every trait gets its own disconnected version. This version would have
+to travel across the api-boundary along with the trait data. Whilst the obvious
+thing would be to enhance the low level `TraitsData` type to have a
+first class understanding of versioning, this cannot be so because not
+every method takes a `TraitsData` as an argument. The lowest-friction
+course of action might be to pack the version into `TraitId`, essentially
+doing string-munging to communicate the version across the api-boundary.
+
+Hosts and managers would need to be inspect the version data, which we
+could provide conveniences for, and construct the correct version
+of the trait view.
+
+```yml
+# traits.yml
+Camera:
+  description: >
+    The entity represents a device that captures a view through
+    re-projection or encoding into some other form.
+  usage:
+    - entity
+  version: 3
+```
+
+```python
+# Manager
+# In this example, CameraTrait is on version 3
+# SpatialTrait is on version 2
+#
+# The manager is responding to a resolve, needing to populate a
+# TraitsData with a camera trait and a spatial trait of the expected
+# version
+cameraVersion = mediacreation.versionFromTraitId(mediacreation.CameraTrait.kId)
+spatialVersion = mediacreation.versionFromTraitId(mediacreation.SpatialTrait.kId)
+
+# Support all known major versions.
+# Camera
+success_result = TraitsData()
+if (cameraVersion >= 3 and cameraVersion < 4):
+     trait = mediacreation.CameraTrait(success_result)
+     ... # Up to date camera population
+elif (cameraVersion >= 2 and cameraVersion < 1):
+     trait = mediacreation.legacy.CameraTrait_v2(success_result)
+     ... # V2 camera population
+elif (cameraVersion >= 1):
+     trait = mediacreation.legacy.CameraTrait_v1(success_result)
+     ... # V1 camera population
```

Comment:
What about an alternate/secondary constructor that accepts a version container. The host prefills this with the versions is understands and is handed one of those versions back?

```
# Ask for versions 0 to 3 and 5. But not 4!
trait = mediacreation.legacy.CameraTrait(success_result, {[0,3},{5,5}])
```

Reply:
:thinking: there is still the problem of the manager knowing what version was used by the host.

Perhaps missing from this example is that the `trait` object has methods on it, which may change depending on version, necessitating the branching.

Reply:
If the version is in the id, then the version the host wants is in the `resolve` request?

Reply:
Is the other challenge that in reality the population of the data will vary from version to version, so you'd still need the conditionals so you can route to the correct code path (eg `trait.setOldName(value)` -> `trait.setNewName(value)`)